Panic in reconciler

```
2023-07-05T09:28:23Z    INFO    controller-runtime.metrics    Metrics server is starting to listen    {"addr": ":8080"}
2023-07-05T09:28:23Z    INFO    setup    Found cert-manager installation in the cluster, features requiring cert-manager are enabled
2023-07-05T09:28:23Z    INFO    setup    Unable to find istio installation in the cluster, features requiring istio are disabled
2023-07-05T09:28:23Z    INFO    setup    Found prometheus operator installation in the cluster, features requiring prometheus operator are enabled
2023-07-05T09:28:23Z    INFO    controller-runtime.builder    Registering a mutating webhook    {"GVK": "temporal.io/v1beta1, Kind=TemporalCluster", "path": "/mutate-temporal-io-v1beta1-temporalcluster"}
2023-07-05T09:28:23Z    INFO    controller-runtime.webhook    Registering webhook    {"path": "/mutate-temporal-io-v1beta1-temporalcluster"}
2023-07-05T09:28:23Z    INFO    controller-runtime.builder    Registering a validating webhook    {"GVK": "temporal.io/v1beta1, Kind=TemporalCluster", "path": "/validate-temporal-io-v1beta1-temporalcluster"}
2023-07-05T09:28:23Z    INFO    controller-runtime.webhook    Registering webhook    {"path": "/validate-temporal-io-v1beta1-temporalcluster"}
2023-07-05T09:28:23Z    INFO    controller-runtime.builder    Registering a mutating webhook    {"GVK": "temporal.io/v1beta1, Kind=TemporalWorkerProcess", "path": "/mutate-temporal-io-v1beta1-temporalworkerprocess"}
2023-07-05T09:28:23Z    INFO    controller-runtime.webhook    Registering webhook    {"path": "/mutate-temporal-io-v1beta1-temporalworkerprocess"}
2023-07-05T09:28:23Z    INFO    controller-runtime.builder    Registering a validating webhook    {"GVK": "temporal.io/v1beta1, Kind=TemporalWorkerProcess", "path": "/validate-temporal-io-v1beta1-temporalworkerprocess"}
2023-07-05T09:28:23Z    INFO    controller-runtime.webhook    Registering webhook    {"path": "/validate-temporal-io-v1beta1-temporalworkerprocess"}
2023-07-05T09:28:23Z    INFO    setup    starting manager
2023-07-05T09:28:23Z    INFO    controller-runtime.webhook.webhooks    Starting webhook server
2023-07-05T09:28:23Z    INFO    controller-runtime.certwatcher    Updated current TLS certificate
2023-07-05T09:28:23Z    INFO    Starting server    {"path": "/metrics", "kind": "metrics", "addr": ":8080"}
2023-07-05T09:28:23Z    INFO    controller-runtime.webhook    Serving webhook server    {"host": "", "port": 9443}
2023-07-05T09:28:23Z    INFO    Starting server    {"kind": "health probe", "addr": ":8081"}
2023-07-05T09:28:23Z    INFO    controller-runtime.certwatcher    Starting certificate watcher
I0705 09:28:23.993985       1 leaderelection.go:248] attempting to acquire leader lease temporal-system/0cfcfa11.temporal.io...
I0705 09:28:40.766997       1 leaderelection.go:258] successfully acquired lease temporal-system/0cfcfa11.temporal.io
2023-07-05T09:28:40Z    INFO    Starting EventSource    {"controller": "temporalcluster", "controllerGroup": "temporal.io", "controllerKind": "TemporalCluster", "source": "kind source: *v1beta1.TemporalCluster"}
2023-07-05T09:28:40Z    INFO    Starting EventSource    {"controller": "temporalcluster", "controllerGroup": "temporal.io", "controllerKind": "TemporalCluster", "source": "kind source: *v1.Deployment"}
2023-07-05T09:28:40Z    INFO    Starting EventSource    {"controller": "temporalcluster", "controllerGroup": "temporal.io", "controllerKind": "TemporalCluster", "source": "kind source: *v1.ConfigMap"}
2023-07-05T09:28:40Z    INFO    Starting EventSource    {"controller": "temporalcluster", "controllerGroup": "temporal.io", "controllerKind": "TemporalCluster", "source": "kind source: *v1.Service"}
2023-07-05T09:28:40Z    INFO    Starting EventSource    {"controller": "temporalcluster", "controllerGroup": "temporal.io", "controllerKind": "TemporalCluster", "source": "kind source: *v1.ServiceAccount"}
2023-07-05T09:28:40Z    INFO    Starting EventSource    {"controller": "temporalcluster", "controllerGroup": "temporal.io", "controllerKind": "TemporalCluster", "source": "kind source: *v1.Ingress"}
2023-07-05T09:28:40Z    INFO    Starting EventSource    {"controller": "temporalcluster", "controllerGroup": "temporal.io", "controllerKind": "TemporalCluster", "source": "kind source: *v1.Job"}
2023-07-05T09:28:40Z    INFO    Starting EventSource    {"controller": "temporalcluster", "controllerGroup": "temporal.io", "controllerKind": "TemporalCluster", "source": "kind source: *v1.Issuer"}
2023-07-05T09:28:40Z    INFO    Starting EventSource    {"controller": "temporalcluster", "controllerGroup": "temporal.io", "controllerKind": "TemporalCluster", "source": "kind source: *v1.Certificate"}
2023-07-05T09:28:40Z    INFO    Starting EventSource    {"controller": "temporalcluster", "controllerGroup": "temporal.io", "controllerKind": "TemporalCluster", "source": "kind source: *v1.ServiceMonitor"}
2023-07-05T09:28:40Z    INFO    Starting Controller    {"controller": "temporalcluster", "controllerGroup": "temporal.io", "controllerKind": "TemporalCluster"}
2023-07-05T09:28:40Z    DEBUG    events    temporal-operator-controller-manager-66d6577656-zf8gq_8536e5fc-4908-48f6-b331-bd7ba1a86693 became leader    {"type": "Normal", "object": {"kind":"Lease","namespace":"temporal-system","name":"0cfcfa11.temporal.io","uid":"3095bf29-99a2-4e35-a661-23570091dab4","apiVersion":"coordination.k8s.io/v1","resourceVersion":"143108137"}, "reason": "LeaderElection"}
2023-07-05T09:28:40Z    INFO    Starting EventSource    {"controller": "temporalclusterclient", "controllerGroup": "temporal.io", "controllerKind": "TemporalClusterClient", "source": "kind source: *v1beta1.TemporalClusterClient"}
2023-07-05T09:28:40Z    INFO    Starting EventSource    {"controller": "temporalclusterclient", "controllerGroup": "temporal.io", "controllerKind": "TemporalClusterClient", "source": "kind source: *v1.Issuer"}
2023-07-05T09:28:40Z    INFO    Starting EventSource    {"controller": "temporalclusterclient", "controllerGroup": "temporal.io", "controllerKind": "TemporalClusterClient", "source": "kind source: *v1.Certificate"}
2023-07-05T09:28:40Z    INFO    Starting EventSource    {"controller": "temporalclusterclient", "controllerGroup": "temporal.io", "controllerKind": "TemporalClusterClient", "source": "kind source: *v1.Secret"}
2023-07-05T09:28:40Z    INFO    Starting Controller    {"controller": "temporalclusterclient", "controllerGroup": "temporal.io", "controllerKind": "TemporalClusterClient"}
2023-07-05T09:28:40Z    INFO    Starting EventSource    {"controller": "temporalworkerprocess", "controllerGroup": "temporal.io", "controllerKind": "TemporalWorkerProcess", "source": "kind source: *v1beta1.TemporalWorkerProcess"}
2023-07-05T09:28:40Z    INFO    Starting EventSource    {"controller": "temporalworkerprocess", "controllerGroup": "temporal.io", "controllerKind": "TemporalWorkerProcess", "source": "kind source: *v1.Deployment"}
2023-07-05T09:28:40Z    INFO    Starting Controller    {"controller": "temporalworkerprocess", "controllerGroup": "temporal.io", "controllerKind": "TemporalWorkerProcess"}
2023-07-05T09:28:40Z    INFO    Starting EventSource    {"controller": "temporalnamespace", "controllerGroup": "temporal.io", "controllerKind": "TemporalNamespace", "source": "kind source: *v1beta1.TemporalNamespace"}
2023-07-05T09:28:40Z    INFO    Starting Controller    {"controller": "temporalnamespace", "controllerGroup": "temporal.io", "controllerKind": "TemporalNamespace"}
2023-07-05T09:28:41Z    INFO    Starting workers    {"controller": "temporalclusterclient", "controllerGroup": "temporal.io", "controllerKind": "TemporalClusterClient", "worker count": 1}
2023-07-05T09:28:41Z    INFO    Starting workers    {"controller": "temporalnamespace", "controllerGroup": "temporal.io", "controllerKind": "TemporalNamespace", "worker count": 1}
2023-07-05T09:28:41Z    INFO    Starting workers    {"controller": "temporalworkerprocess", "controllerGroup": "temporal.io", "controllerKind": "TemporalWorkerProcess", "worker count": 1}
2023-07-05T09:28:41Z    INFO    Starting workers    {"controller": "temporalcluster", "controllerGroup": "temporal.io", "controllerKind": "TemporalCluster", "worker count": 1}
2023-07-05T09:28:41Z    INFO    Starting reconciliation    {"controller": "temporalcluster", "controllerGroup": "temporal.io", "controllerKind": "TemporalCluster", "TemporalCluster": {"name":"temporal-cluster","namespace":"dgt-system"}, "namespace": "dgt-system", "name": "temporal-cluster", "reconcileID": "1cd7bde1-e8a5-4d6a-ab24-b4bdf831d8b4"}
2023-07-05T09:28:41Z    INFO    Starting reconciliation    {"controller": "temporalnamespace", "controllerGroup": "temporal.io", "controllerKind": "TemporalNamespace", "TemporalNamespace": {"name":"default","namespace":"dgt-system"}, "namespace": "dgt-system", "name": "default", "reconcileID": "ed7d8ea7-052a-41fb-9a30-68aca1780ed2"}
2023-07-05T09:28:41Z    INFO    Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference    {"controller": "temporalnamespace", "controllerGroup": "temporal.io", "controllerKind": "TemporalNamespace", "TemporalNamespace": {"name":"default","namespace":"dgt-system"}, "namespace": "dgt-system", "name": "default", "reconcileID": "ed7d8ea7-052a-41fb-9a30-68aca1780ed2"}
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
    panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x15278ac]

goroutine 529 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
    /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:119 +0x1fa
panic({0x201a0e0, 0x3aaef30})
    /usr/local/go/src/runtime/panic.go:884 +0x212
github.com/alexandrevilain/temporal-operator/api/v1beta1.(*TemporalCluster).GetPublicClientAddress(0xc000fc6000)
    /workspace/api/v1beta1/temporalcluster_types.go:838 +0x10c
github.com/alexandrevilain/temporal-operator/pkg/temporal.buildClusterClientOptions({_, _}, {_, _}, _, {_, _, _})
    /workspace/pkg/temporal/client.go:95 +0x97
github.com/alexandrevilain/temporal-operator/pkg/temporal.GetClusterNamespaceClient({0x2881618, 0xc000af4a50}, {0x288c108?, 0xc00044a660?}, 0x2881618?, {0x0?, 0xc0004ef480?, 0x0?})
    /workspace/pkg/temporal/client.go:148 +0x85
github.com/alexandrevilain/temporal-operator/controllers.(*TemporalNamespaceReconciler).Reconcile(0xc0005bd278, {0x2881618, 0xc000af4a50}, {{{0xc0004ef480?, 0x10?}, {0xc0004ef470?, 0x40da87?}}})
    /workspace/controllers/temporalnamespace_controller.go:107 +0x448
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x2881618?, {0x2881618?, 0xc000af4a50?}, {{{0xc0004ef480?, 0x1f710a0?}, {0xc0004ef470?, 0x0?}}})
    /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:122 +0xc8
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0002d00a0, {0x2881570, 0xc00080c1c0}, {0x20babe0?, 0xc000050280?})
    /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:323 +0x3a5
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0002d00a0, {0x2881570, 0xc00080c1c0})
    /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:274 +0x1d9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
    /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:235 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
    /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:231 +0x333
Stream closed EOF for temporal-system/temporal-operator-controller-manager-66d6577656-zf8gq (manager)

```

--- a/api/v1beta1/temporalcluster_types.go
+++ b/api/v1beta1/temporalcluster_types.go
@@ -871,6 +871,16 @@ func (c *TemporalCluster) GetPublicClientAddress() string {
 	return fmt.Sprintf("%s.%s:%d", c.ChildResourceName("frontend"), c.GetNamespace(), *c.Spec.Services.Frontend.Port)
 }
 
+// IsReady returns true if the TemporalCluster's conditions reports it ready.
+func (c *TemporalCluster) IsReady() bool {
+	for _, condition := range c.Status.Conditions {
+		if condition.Type == ReadyCondition && condition.Status == metav1.ConditionTrue {
+			return true
+		}
+	}
+	return false
+}
+
 //+kubebuilder:object:root=true
 
 // TemporalClusterList contains a list of Cluster.

--- a/api/v1beta1/temporalnamespace_types.go
+++ b/api/v1beta1/temporalnamespace_types.go
@@ -18,14 +18,13 @@
 package v1beta1
 
 import (
-	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
 // TemporalNamespaceSpec defines the desired state of Namespace.
 type TemporalNamespaceSpec struct {
 	// Reference to the temporal cluster the namespace will be created.
-	ClusterRef corev1.LocalObjectReference `json:"clusterRef"`
+	ClusterRef TemporalClusterReference `json:"clusterRef"`
 	// Namespace description.
 	// +optional
 	Description string `json:"description,omitempty"`

--- a/controllers/temporalclusterclient_controller.go
+++ b/controllers/temporalclusterclient_controller.go
@@ -93,6 +93,12 @@ func (r *TemporalClusterClientReconciler) Reconcile(ctx context.Context, req ctr
 		return reconcile.Result{}, err
 	}
 
+	if !cluster.IsReady() {
+		logger.Info("Skipping cluster client reconciliation until referenced cluster is ready")
+
+		return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
+	}
+
 	if !(cluster.MTLSWithCertManagerEnabled() && cluster.Spec.MTLS.FrontendEnabled()) {
 		return reconcile.Result{Requeue: false}, errors.New("mTLS for frontend not enabled using cert-manager for the cluster, can't create a client")
 	}

--- a/controllers/temporalnamespace_controller.go
+++ b/controllers/temporalnamespace_controller.go
@@ -28,7 +28,6 @@ import (
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
-	"k8s.io/apimachinery/pkg/types"
 	kerrors "k8s.io/apimachinery/pkg/util/errors"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/builder"
@@ -70,11 +69,16 @@ func (r *TemporalNamespaceReconciler) Reconcile(ctx context.Context, req ctrl.Re
 		return reconcile.Result{}, err
 	}
 
-	namespacedName := types.NamespacedName{Namespace: req.Namespace, Name: namespace.Spec.ClusterRef.Name}
 	cluster := &v1beta1.TemporalCluster{}
-	err = r.Get(ctx, namespacedName, cluster)
+	err = r.Get(ctx, namespace.Spec.ClusterRef.NamespacedName(namespace), cluster)
 	if err != nil {
 		return r.handleError(ctx, namespace, v1beta1.ReconcileErrorReason, err)
+	}
+
+	if !cluster.IsReady() {
+		logger.Info("Skipping namespace reconciliation until referenced cluster is ready")
+
+		return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
 	}
 
 	patchHelper, err := patch.NewHelper(namespace, r.Client)

--- a/controllers/temporalworkerprocess_controller.go
+++ b/controllers/temporalworkerprocess_controller.go
@@ -107,6 +107,12 @@ func (r *TemporalWorkerProcessReconciler) Reconcile(ctx context.Context, req ctr
 		return r.handleError(ctx, worker, v1beta1.ReconcileErrorReason, err)
 	}
 
+	if !cluster.IsReady() {
+		logger.Info("Skipping workerprocess reconciliation until referenced cluster is ready")
+
+		return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
+	}
+
 	if requeueAfter, err := r.reconcileResources(ctx, worker, cluster); err != nil || requeueAfter > 0 {
 		if err != nil {
 			logger.Error(err, "Can't reconcile resources")

--- a/tests/e2e/assert_test.go
+++ b/tests/e2e/assert_test.go
@@ -231,7 +231,7 @@ func AssertCanCreateTemporalNamespace(name string) features.Func {
 		temporalNamespace := &v1beta1.TemporalNamespace{
 			ObjectMeta: metav1.ObjectMeta{Name: name, Namespace: namespace},
 			Spec: v1beta1.TemporalNamespaceSpec{
-				ClusterRef: corev1.LocalObjectReference{
+				ClusterRef: v1beta1.TemporalClusterReference{
 					Name: cluster.GetName(),
 				},
 				RetentionPeriod: &metav1.Duration{Duration: 24 * time.Hour},

--- a/tests/e2e/namespace_test.go
+++ b/tests/e2e/namespace_test.go
@@ -26,7 +26,6 @@ import (
 	"github.com/alexandrevilain/temporal-operator/api/v1beta1"
 	"github.com/alexandrevilain/temporal-operator/pkg/temporal"
 	"go.temporal.io/api/serviceerror"
-	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/util/retry"
 
@@ -59,7 +58,7 @@ func TestNamespaceCreation(t *testing.T) {
 			temporalNamespace = &v1beta1.TemporalNamespace{
 				ObjectMeta: metav1.ObjectMeta{Name: "test", Namespace: namespace},
 				Spec: v1beta1.TemporalNamespaceSpec{
-					ClusterRef: corev1.LocalObjectReference{
+					ClusterRef: v1beta1.TemporalClusterReference{
 						Name: cluster.GetName(),
 					},
 					RetentionPeriod: &metav1.Duration{Duration: 24 * time.Hour},

--- a/tests/e2e/utils_test.go
+++ b/tests/e2e/utils_test.go
@@ -191,12 +191,7 @@ func waitForDeployment(ctx context.Context, cfg *envconf.Config, dep *appsv1.Dep
 // waitForCluster waits for the temporal cluster's components to be up and running (reporting Ready condition).
 func waitForCluster(ctx context.Context, cfg *envconf.Config, cluster *v1beta1.TemporalCluster) error {
 	cond := conditions.New(cfg.Client().Resources()).ResourceMatch(cluster, func(object k8s.Object) bool {
-		for _, condition := range object.(*v1beta1.TemporalCluster).Status.Conditions {
-			if condition.Type == v1beta1.ReadyCondition && condition.Status == metav1.ConditionTrue {
-				return true
-			}
-		}
-		return false
+		return object.(*v1beta1.TemporalCluster).IsReady()
 	})
 	return wait.For(cond, wait.WithTimeout(time.Minute*10))
 }

